File: src/vehicle_controller/src/msc_if.cpp
#include "../include/vehicle_controller/msc_if.h"

#include <fcntl.h>    //access the fd
#include <termios.h>  //for serial commuication options
#include <unistd.h>   //for closing the file descriptor

#include <cstring>
#include <rclcpp/logging.hpp>  //for logging write error
#include <stdexcept>
#include <string>

// adding
using std::runtime_error;
using std::to_string;

msc_if::msc_if(const char *usb_path) {
    this->fd = open(usb_path, O_RDWR | O_NOCTTY);
    if (fd == -1) {
        RCLCPP_ERROR(get_logger(), "Failed to open servo USB (%s)", usb_path);
        throw runtime_error("failed to open usb path to Servo Controller");
    }

    tcgetattr(this->fd, &(this->options));
    this->options.c_iflag &= ~(INLCR | IGNCR | ICRNL | IXON | IXOFF);
    this->options.c_oflag &= ~(ONLCR | OCRNL);
    this->options.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
    tcsetattr(this->fd, TCSANOW, &(this->options));

    RCLCPP_INFO(get_logger(),
                "Maestro USB Servo Controller has successfully initalized "
                "using %s path.",
                usb_path);
}

msc_if::~msc_if() { close(this->fd); }

void msc_if::set_servo(const unsigned char servo_pin, const unsigned short servo_value) {
    // MSC expects servo value in units of quarter of usecs
    unsigned short converted = servo_value * 4;
    unsigned char command[] = {0x84, servo_pin,
                               static_cast<unsigned char>(converted & 0x7F),
                               static_cast<unsigned char>(converted >> 7 & 0x7F)};

    RCLCPP_DEBUG(get_logger(), "Command: %10x %10x %10x %10x \n", command[0], command[1],
                 command[2], command[3]);
    RCLCPP_DEBUG(get_logger(), "Setting servo %d (%d us).\n", converted, servo_value);
    int bytes_send = write(this->fd, command, sizeof(command));

    if (bytes_send == -1) {
        // failed to send command
        int error = errno;
        RCLCPP_ERROR(get_logger(), "Failed to send command %s (errno += %d, fd = %d)",
                     strerror(error), error, this->fd);

        throw runtime_error("Failed to send command: %s");
    }
}


void msc_if::set_pin(const unsigned char gpio_pin, int value){
      unsigned short servo_value;  // Add this declaration

      if(value == 0){
          servo_value = 0;
          RCLCPP_DEBUG(get_logger(), "Turning motor pin %d off", gpio_pin);
      } else{
          //any value over 6000 will equal logical high on the servo pin
          servo_value = 7000;
          RCLCPP_DEBUG(get_logger(), "Turning motor pin %d on", gpio_pin);

      }

      unsigned short converted = servo_value;
      unsigned char command[] = {0x84, gpio_pin,  
                                 static_cast<unsigned char>(converted & 0x7F),
                                 static_cast<unsigned char>(converted >> 7 & 0x7F)};

    RCLCPP_DEBUG(get_logger(), "Command: %10x %10x %10x %10x \n", command[0], command[1],
                 command[2], command[3]);
    RCLCPP_DEBUG(get_logger(), "Setting servo %d (%d us).\n", converted, servo_value);
    int bytes_send = write(this->fd, command, sizeof(command));

    if (bytes_send == -1) {
        // failed to send command
        int error = errno;
        RCLCPP_ERROR(get_logger(), "Failed to send command %s (errno += %d, fd = %d)",
                     strerror(error), error, this->fd);

        throw runtime_error("Failed to send command: %s");
    }
}

File: src/vehicle_controller_py/vehicle_controller_py/msc_if.py
import serial
import rclpy

class MscIf:
    def __init__(self, usb_path: str, logger=None):
        self.usb_path = usb_path
        self.logger = logger or rclpy.logging.get_logger("maestro_servo_controller")
        
        try:
            # pyserial automatically handles the raw termios flag configuration
            self.serial_port = serial.Serial(self.usb_path, timeout=1.0)
            self.logger.info(f"Maestro USB Servo Controller has successfully initialized using {self.usb_path} path.")
        except serial.SerialException as e:
            self.logger.error(f"Failed to open servo USB ({self.usb_path}): {e}")
            raise RuntimeError(f"failed to open usb path to Servo Controller: {e}")

    def __del__(self):
        """Clean up the serial port when the object is destroyed."""
        if hasattr(self, 'serial_port') and self.serial_port.is_open:
            self.serial_port.close()

    def set_servo(self, servo_pin: int, servo_value: int):
        """
        Sets the position of a standard servo.
        The Pololu MSC expects target values in units of quarter-microseconds.
        """
        converted = int(servo_value * 4)
        
        # Protocol: 0x84, channel, lower 7 bits, upper 7 bits
        command = bytearray([
            0x84,
            servo_pin,
            converted & 0x7F,
            (converted >> 7) & 0x7F
        ])

        self.logger.debug(
            f"Command: {command[0]:02x} {command[1]:02x} {command[2]:02x} {command[3]:02x} "
            f"| Setting servo {servo_pin} ({servo_value} us)."
        )
        
        try:
            self.serial_port.write(command)
        except serial.SerialException as e:
            self.logger.error(f"Failed to send command: {e}")
            raise RuntimeError(f"Failed to send command: {e}")

    def set_pin(self, gpio_pin: int, value: int):
        """
        Sets a pin as a digital output high or low.
        """
        if value == 0:
            servo_value = 0
            self.logger.debug(f"Turning motor pin {gpio_pin} off")
        else:
            # Any value over 6000 will equal logical high on the servo pin
            servo_value = 7000
            self.logger.debug(f"Turning motor pin {gpio_pin} on")

        # Create bytearray command without the *4 conversion
        command = bytearray([
            0x84,
            gpio_pin,
            servo_value & 0x7F,
            (servo_value >> 7) & 0x7F
        ])

        self.logger.debug(
            f"Command: {command[0]:02x} {command[1]:02x} {command[2]:02x} {command[3]:02x} "
            f"| Setting GPIO pin {gpio_pin} to logical state {value}."
        )
        
        try:
            self.serial_port.write(command)
        except serial.SerialException as e:
            self.logger.error(f"Failed to send command: {e}")
            raise RuntimeError(f"Failed to send command: {e}")